### ISSUE TITLE
Universal link not working

### ISSUE THREAD

**Issue by DamienDulac** [adjust/react_native_sdk#242]
Hi,

I am taking the liberty of creating an issue because I cannot process the deep link which is passed as a parameter of a universal link on my react native iOS app. The link is: **https://myprojectid.adj.st/promotions?adj_t=........&adj_deep_link=fidmelistint%3A%2F%2Ffidmelist%2Fpromotions**.

For several days I have been trying to find out why the link opens my application, but redirects me each time to the iOS store.

I tried following the documentation (https://dev.adjust.com/en/sdk/ios/features/deep-links/direct?version=v4#:~:text=your%20app%20delegate.-,Universal%20links,-Update%20the%20application), but even by implementing my own logic to retrieve the deep link via the adj_deep_link parameter present in the URL, I am automatically redirected to the store.

Do you have any sample code that we are supposed to use to make universal linking work correctly?

` 
        [ADJLinkResolution

        resolveLinkWithUrl:incomingURL

        resolveUrlSuffixArray:@[

        @"email.example.com", @"short.example.com"

        ]
       
        callback:^(NSURL* _Nullable resolvedURL) {

            // add your code below to handle deep link

            // (for example, open deep link content)

            // resolvedURL object contains the deep link

            // call the below method to send deep link to Adjust's servers

            [Adjust appWillOpenUrl:resolvedURL];
        }];`

How does the SDK know that the application was opened via the deep link, and that it should not be redirected to the appstore? And is it normal that the content of resolvedURL is the same as incomingURL ?

Thank you in advance for your help.

**Comment by marcos-cunha-gringo** [comment]
> Hi,
> 
> I am taking the liberty of creating an issue because I cannot process the deep link which is passed as a parameter of a universal link on my react native iOS app. The link is: **https://myprojectid.adj.st/promotions?adj_t=........&adj_deep_link=fidmelistint%3A%2F%2Ffidmelist%2Fpromotions**.
> 
> For several days I have been trying to find out why the link opens my application, but redirects me each time to the iOS store.
> 
> I tried following the documentation (https://dev.adjust.com/en/sdk/ios/features/deep-links/direct?version=v4#:~:text=your%20app%20delegate.-,Universal%20links,-Update%20the%20application), but even by implementing my own logic to retrieve the deep link via the adj_deep_link parameter present in the URL, I am automatically redirected to the store.
> 
> Do you have any sample code that we are supposed to use to make universal linking work correctly?
> 
> ` [ADJLinkResolution
> 
> ```
>     resolveLinkWithUrl:incomingURL
> 
>     resolveUrlSuffixArray:@[
> 
>     @"email.example.com", @"short.example.com"
> 
>     ]
>    
>     callback:^(NSURL* _Nullable resolvedURL) {
> 
>         // add your code below to handle deep link
> 
>         // (for example, open deep link content)
> 
>         // resolvedURL object contains the deep link
> 
>         // call the below method to send deep link to Adjust's servers
> 
>         [Adjust appWillOpenUrl:resolvedURL];
>     }];`
> ```
> 
> How does the SDK know that the application was opened via the deep link, and that it should not be redirected to the appstore? And is it normal that the content of resolvedURL is the same as incomingURL ?
> 
> Thank you in advance for your help.

Hi @DamienDulac! I was searching for a solution because at the company where I work, we are experiencing the same problem, but only with iOS. On Android, the library works as expected. Which version of React Native are you using? And which Adjust SDK version?

We're using:

RN: 0.74.4
Adjust SDK Version: v4

**Comment by DamienDulac** [comment]
Hi @marcos-cunha-gringo, 

On our side we also encounter the problem on iOS, and everything works fine on android. We are on react-native 0.71.19 and adjust 4.29.6.

Have you managed to solve this problem since then?

Thanks in advance.

**Comment by jamesboone** [comment]
Can you try upgrading to the Adjust SDK v5, there are important updates to this SDK that might resolve your issues.

**Comment by uerceg** [comment]
Hi @DamienDulac,

Sorry for the delay in replying and @jamesboone, thank you for jumping in and the suggestion.

> I cannot process the deep link which is passed as a parameter of a universal link on my react native iOS app. The link is: https://myprojectid.adj.st/promotions?adj_t=........&adj_deep_link=fidmelistint%3A%2F%2Ffidmelist%2Fpromotions.

Are you trying to use this universal link to directly open your already installed app?

> I tried following the documentation (https://dev.adjust.com/en/sdk/ios/features/deep-links/direct?version=v4#:~:text=your%20app%20delegate.-,Universal%20links,-Update%20the%20application), but even by implementing my own logic to retrieve the deep link via the adj_deep_link parameter present in the URL, I am automatically redirected to the store.

If the answer to the previous question is yes, this looks like a good place to look for how to set up universal links in your iOS app. I'm still failing to understand - is that your app is not being opened with the universal link at all or it is being opened, but you are unable to parse the link to get the `adj_deep_link` parameter from it?

> How does the SDK know that the application was opened via the deep link

The SDK does not know this, it's the knowledge of the operating system and the app API. Natively, there are certain methods in the `AppDelegate` which will get triggered when the deep link opens an app. Our SDK has no access to those methods and inside of our guides we are saying to people that they should obtain the link that is opening their app and pass this link to us via (in v5 terms) `processDeeplink` method (or `processAndResolveDeeplink` in case short link is used and one would like to see the unshortened version of the link).

> and that it should not be redirected to the appstore?

This is not something that the SDK does, but our backend instead. When you tap on the universal link, our backend will try to understand in "chat" with operating system whether app is installed and will be opened or should it redirect to the App Store.

> And is it normal that the content of resolvedURL is the same as incomingURL ?

If you are referring to `resolveLinkWithUrl:resolveUrlSuffixArray:callback:` method, in theory, yes. Purpose of that method is to take the URL you're giving to it and to try to follow URL redirects until it either finds (1) any domain from the ones you have passed as URL suffix array or (2) if it finds [some of the Adjust domains](https://github.com/adjust/ios_sdk/blob/master/Adjust/ADJLinkResolution.m#L180). So in theory, if you would send a link starting with `https://myprojectid.adj.st` to that method to be resolved, it would be immediately echoed back since it's already link that that method thinks as one that shouldn't be resolved any further.

This method is being used by clients who are having Adjust URLs wrapped into some other URLs (email campaigns, own shorteners, etc). Then this method can help to locate the underlying Adjust link from the one that wraps it.

---

Having in mind that the question has been posted a while ago, I hope that you have found the answers to your questions already in the meantime and I'll close this issue. However, if you still have some questions and need our help with anything, please feel free to comment / reopen.

Cheers.

**Comment by mzaien** [comment]
I'm on sdk 5 and still having the same issue, added associated links, have setup linking to check the domains 
clean build, release mode I checked every possible way to resolve this issue

it work on android as expected but on ios we are failing to nail the last part ...

if you need any info about my setup to help me fix it I would love to give and help
@uerceg

**Comment by uerceg** [comment]
Hey @mzaien,

Can you quickly explain your setup and how does your test look like (what are the steps you're performing, what is that you're expecting to happen, what is happening instead)?

**Comment by mzaien** [comment]
Hello @uerceg, I'm running an cng expo app with adjust, I added the needed configs, associatedDomains set up linking with expo-linking and react navigation 

when I have the app installed either app store or direct install, clicking adjust link always opens the app store instead of the app

```ts
associatedDomains: ['applinks:pmlw.adj.st', 'applinks:malaa.go.link'],
```

also tried the reversed order 

```ts
associatedDomains: ['applinks:malaa.go.link', 'applinks:pmlw.adj.st'],
```

<img width="977" height="147" alt="Image" src="https://github.com/user-attachments/assets/c1dc8504-33cd-49a5-b3ec-780c50a36d32" />

https://github.com/user-attachments/assets/ea010142-4241-475c-bf0b-a5b4701fd9d3


I'm using the latest adjust sdk 5.4.4 with expo 54 and react native 0.81.2 old arch

**Comment by uerceg** [comment]
Thanks for the detailed explanation. Looping in @karthik-kannan-adjust to take a look and help with this one.

**Comment by uerceg** [comment]
@mzaien Quick question before digging deeper on our end: In your video, once you have been redirected to the App Store for your app, one can see that there's a button for your app to be downloaded and installed on the device. Whereas if the app was installed on the device, there would be an option to tap the button "Open".

Is the app you're trying to open with this deep link installed on that device?

**Comment by mzaien** [comment]
@uerceg yes I installed it manually with `yarn expo run:ios --configuration Release` to test adding the `applinks:pmlw.adj.st`
to associated domains

**Comment by uerceg** [comment]
Just to double check - is the bundle ID of the build that gets generated and installed with this `yarn expo run` command the same as the one from the app you have released in the App Store?

**Comment by mzaien** [comment]
> Just to double check - is the bundle ID of the build that gets generated and installed with this `yarn expo run` command the same as the one from the app you have released in the App Store?

Yes same bundle id which is the same one in the AASA hosted by adjust servers

**Comment by karthik-kannan-adjust** [comment]
Hey @mzaien Nice to connect. Looking into your issue

Seems like a classic case of iOS UL not working because of some minor issue.

From my side, I see that

- AASA file: https://malaa.go.link/.well-known/apple-app-site-association. Looks to be set up correctly and hosted
- Bundle ID: tech.malaa.personal and App ID prefix : X4J2ZQTH86. Looks to be entered correctly in Adjust Dash
- Associated domains: associatedDomains: ['applinks:pmlw.adj.st', 'applinks:malaa.go.link']. Looks to be configured correctly

So the domain malaa.go.link should ideally open the app and not redirect you to store if everything else lines up and iOS UL is working correctly.

Can you please try the step on https://developer.apple.com/documentation/technotes/tn3155-debugging-universal-links#Test-universal-links-behavior i.e., paste  https://malaa.go.link/?adj_t=1ug18lov in your iOS notes app, long press and and see if you get the option to open in app

`To test your universal links behavior, paste a link into your Notes app and long-press it (iOS) or control-click it (macOS) to see your options for following the link. If universal links have been configured correctly, the option to open in your app and in the web browser will both show up. The option you choose will set the default behavior for your device when following universal links from this domain in the future. To change this default choice, repeat the same steps and choose a different option.`

Also worth testing on a different device, as sometimes UL has some quirks where if you chose to open in browser once, it persists that setting and always opens in browser rather than attempting to open the app.

**Comment by mzaien** [comment]
Hello 👋 @karthik-kannan-adjust 
Here is a demo from two device 

https://github.com/user-attachments/assets/5ca2490d-3321-44b9-b548-7e96316ac854

https://github.com/user-attachments/assets/0b0be1c6-85ee-4d99-b8ea-237556861cac

**Comment by karthik-kannan-adjust** [comment]
@mzaien Got it, so in both devices, we don't have the option to "open in app" when the link is long-pressed in iOS notes app. This basically confirms that some issue/something is missing with the Universal links setup.

Can you release a version of the app to the App Store with the domains `malaa.go.link` and `pmlw.adj.st` added as associated domains? I suppose if those are the only changes, there is little risk of anything else breaking on the live version?

In parallel, I will check in with our internal teams to see if they can find anything obviously wrong with the UL setup. Will reply in this thread if I find anything.

**Comment by karthik-kannan-adjust** [comment]
@mzaien I checked again, and don't see anything obviously wrong with the overall UL setup/config.

- (Preferred) If you manage to release a version of your app with associated domains added to the App Store, let us know, and we can test it
- If not, we would need access to your TestFlight build OR the IPA file to check further. For this, rather than handling it in the GitHub issue here, it would be better if you start a ticket with our support team support@adjust.com, so that they can dive deeper and check the UL setup. Thanks for understanding!
